cpp/src/arrow/types/struct.h

```diff
@@ -23,7 +23,137 @@
 #include <vector>
 
 #include "arrow/type.h"
+#include "arrow/types/primitive.h"
+#include "arrow/types/list.h"
 
-namespace arrow {}  // namespace arrow
+namespace arrow {
 
-#endif  // ARROW_TYPES_STRUCT_H
+class StructArray : public Array {
+ public:
+  StructArray(const TypePtr& type, int32_t length,
+      std::vector<ArrayPtr>& values,
+      int32_t null_count = 0,
+      std::shared_ptr<Buffer> null_bitmap = nullptr) :
+      Array(type, length, null_count, null_bitmap) {
+    type_ = type;
+    values_ = values;
+  }
+
```

Comment:
I wonder if it's easy to move kinds of `Validate`, like `Validate`for `ListBuilder` or others, to an independent `.h` file, so the programs can use it flexibly.